### ISSUE TITLE
[SLURM]NoSuchFileException

### ISSUE THREAD

**Issue by ScriBanana** [openmole/openmole#514]
Hi everyone (again 😬),

I'm a bit struggling with OM and SLURM and I am unable to figure out what I did wrong when calling `SLURMEnvironment`.

_Setup :_
Here's a little test script I use to explore the issue. It does run properly on `LocalEnvironment` as well as on `SSHEnvironment` on my head node on the cluster, so I reckon it definitely is an issue related to SLURM. I built from 1281cba9ac3fb57b2b9eba44172937e74e25978c, so as up to date as can be.

<details>
<summary>OMS script </summary>

```scala
val i = Val[Double]
val res = Val[Double]

val model =
  ScalaTask("val res = i * 2") set (
    inputs += i,
    outputs += (i, res)
)

val env =
  SLURMEnvironment(
    "<user>",
    "<server>",
    partition = "agap_short",
    name = "OMtest",
    memory = 100 megabytes,
    nodes = 1,
    time = 30 minutes,
    workDirectory = "~/scratch"
  )

DirectSampling(
  evaluation = model on env by 100 hook display,
  sampling = i in (0.0 to 1000.0 by 1.0)
)
```
</details>

_Description :_
Jobs are submitted to SLURM, get pending, and fail immediately as they reach the top of the queue and get assigned a node. OpenMOLE then attempts to submit new jobs again indefinitely. It returns a `No such file` as well as `Job status is FAILED` error. Jobs are indeed _FAILED_ server-side, with a `NonZeroExitCode` reason.

<details>
<summary>"Job status is FAILED" error log </summary>

```batchfile
Job status is FAILED
org.openmole.core.exception.InternalProcessingError: Job status is FAILED
	at org.openmole.plugin.environment.batch.refresh.RefreshActor$$anonfun$receive$1.apply(RefreshActor.scala:40)
	at org.openmole.plugin.environment.batch.refresh.JobManager$.killOr(JobManager.scala:109)
	at org.openmole.plugin.environment.batch.refresh.RefreshActor$.receive(RefreshActor.scala:33)
	at org.openmole.plugin.environment.batch.refresh.JobManager$DispatcherActor$.receive(JobManager.scala:54)
	at org.openmole.plugin.environment.batch.refresh.JobManager$$anonfun$dispatch$1.apply$mcV$sp(JobManager.scala:63)
	at org.openmole.core.threadprovider.ThreadProvider$RunClosure.run(ThreadProvider.scala:24)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
</details>

<details>
<summary>"No such file" error log </summary>

```batchfile
No such file
gridscale.ssh.sshj.SSHClient$NoSuchFileException: No such file
	at gridscale.ssh.sshj.SSHClient$$anon$2.wrap(SSHClient.scala:129)
	at gridscale.ssh.sshj.SSHClient$$anon$2.$anonfun$withClient$1(SSHClient.scala:138)
	at scala.util.Success.flatMap(Try.scala:258)
	at gridscale.ssh.sshj.SSHClient$$anon$2.withClient(SSHClient.scala:137)
	at gridscale.ssh.sshj.SSHClient$$anon$2.rm(SSHClient.scala:163)
	at gridscale.ssh.package$.$anonfun$rmFile$1(package.scala:333)
	at gridscale.ssh.sshj.SSHClient$.withSFTP(SSHClient.scala:48)
	at gridscale.ssh.package$SSH.$anonfun$withSFTP$1(package.scala:90)
	at gridscale.ssh.package$SSHCache$.withCache(package.scala:72)
	at gridscale.ssh.package$SSH.withSFTP(package.scala:89)
	at gridscale.ssh.package$.rmFile(package.scala:333)
	at gridscale.cluster.HeadNode$$anon$1.rmFile(HeadNode.scala:17)
	at gridscale.cluster.HeadNode$$anon$1.rmFile(HeadNode.scala:13)
	at gridscale.cluster.BatchScheduler$.clean(BatchScheduler.scala:76)
	at gridscale.slurm.package$.clean(package.scala:148)
	at org.openmole.plugin.environment.slurm.SLURMJobService$$anonfun$delete$1.apply$mcV$sp(SLURMJobService.scala:63)
	at org.openmole.plugin.environment.slurm.SLURMJobService$$anonfun$delete$1.apply(SLURMJobService.scala:63)
	at org.openmole.plugin.environment.slurm.SLURMJobService$$anonfun$delete$1.apply(SLURMJobService.scala:63)
	at org.openmole.plugin.environment.batch.environment.AccessControl$.withPermit(AccessControl.scala:17)
	at org.openmole.plugin.environment.batch.environment.AccessControl.apply(AccessControl.scala:48)
	at org.openmole.plugin.environment.slurm.SLURMJobService.delete(SLURMJobService.scala:63)
	at org.openmole.plugin.environment.slurm.SLURMEnvironment$$anonfun$submit$3.apply(SLURMEnvironment.scala:146)
	at org.openmole.plugin.environment.slurm.SLURMEnvironment$$anonfun$submit$3.apply(SLURMEnvironment.scala:146)
	at org.openmole.plugin.environment.ssh.package$$anonfun$submitToCluster$2$$anonfun$apply$7.apply$mcV$sp(package.scala:209)
	at org.openmole.plugin.environment.ssh.package$$anonfun$submitToCluster$2$$anonfun$apply$7.apply(package.scala:209)
	at org.openmole.plugin.environment.ssh.package$$anonfun$submitToCluster$2$$anonfun$apply$7.apply(package.scala:209)
	at org.openmole.core.tools.service.Retry$.retry(Retry.scala:37)
	at org.openmole.core.tools.service.Retry$.retry(Retry.scala:47)
	at org.openmole.plugin.environment.batch.refresh.KillActor$.org$openmole$plugin$environment$batch$refresh$KillActor$$kill$1(KillActor.scala:46)
	at org.openmole.plugin.environment.batch.refresh.KillActor$$anonfun$tryKillAndClean$1.apply(KillActor.scala:49)
	at org.openmole.plugin.environment.batch.refresh.KillActor$$anonfun$tryKillAndClean$1.apply(KillActor.scala:49)
	at scala.Option.foreach(Option.scala:437)
	at org.openmole.plugin.environment.batch.refresh.KillActor$.tryKillAndClean(KillActor.scala:49)
	at org.openmole.plugin.environment.batch.refresh.KillActor$.receive(KillActor.scala:40)
	at org.openmole.plugin.environment.batch.refresh.JobManager$DispatcherActor$.receive(JobManager.scala:58)
	at org.openmole.plugin.environment.batch.refresh.JobManager$$anonfun$dispatch$1.apply$mcV$sp(JobManager.scala:63)
	at org.openmole.core.threadprovider.ThreadProvider$RunClosure.run(ThreadProvider.scala:24)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
</details>

<details>
<summary>"scontrol" log on one failed job </summary>

```batchfile
JobId=5636368 JobName=job-c60603ea-374e-4025-b61f-2f42ccad4c7d.slurm
   UserId=<user> GroupId=agap(1082) MCS_label=N/A
   Priority=4294731579 Nice=0 Account=agap QOS=qos_agap
   JobState=FAILED Reason=NonZeroExitCode Dependency=(null)
   Requeue=0 Restarts=0 BatchFlag=1 Reboot=0 ExitCode=1:0
   RunTime=00:00:00 TimeLimit=00:30:00 TimeMin=N/A
   SubmitTime=2021-12-09T18:00:10 EligibleTime=2021-12-09T18:00:10
   StartTime=2021-12-09T18:00:35 EndTime=2021-12-09T18:00:35 Deadline=N/A
   PreemptTime=None SuspendTime=None SecsPreSuspend=0
   Partition=agap_short AllocNode:Sid=muse-login02:194463
   ReqNodeList=(null) ExcNodeList=(null)
   NodeList=muse205
   BatchHost=muse205
   NumNodes=1 NumCPUs=1 NumTasks=0 CPUs/Task=1 ReqB:S:C:T=0:0:*:*
   TRES=cpu=1,mem=100M,node=1
   Socks/Node=* NtasksPerN:B:S:C=0:0:*:* CoreSpec=*
   MinCPUsNode=1 MinMemoryNode=100M MinTmpDiskNode=0
   Features=(null) Gres=(null) Reservation=(null)
   OverSubscribe=OK Contiguous=0 Licenses=(null) Network=(null)
   Command=/nfs/home/<user>/.openmole/.tmp/ssh/openmole-4d13481e-7466-4fb8-815d-e20d0f02b368/tmp/1639065600000/1639069194045_3632f8e5-f092-4696-b377-43c5f5644433/job-c60603ea-374e-4025-b61f-2f42ccad4c7d.slurm
   WorkDir=/nfs/home/<user>/.openmole/.tmp/ssh/openmole-4d13481e-7466-4fb8-815d-e20d0f02b368/tmp/1639065600000/1639069194045_3632f8e5-f092-4696-b377-43c5f5644433
   StdErr=/nfs/home/<user>/.openmole/.tmp/ssh/openmole-4d13481e-7466-4fb8-815d-e20d0f02b368/tmp/1639065600000/1639069194045_3632f8e5-f092-4696-b377-43c5f5644433/job-c60603ea-374e-4025-b61f-2f42ccad4c7d.err
   StdIn=/dev/null
   StdOut=/nfs/home/<user>/.openmole/.tmp/ssh/openmole-4d13481e-7466-4fb8-815d-e20d0f02b368/tmp/1639065600000/1639069194045_3632f8e5-f092-4696-b377-43c5f5644433/job-c60603ea-374e-4025-b61f-2f42ccad4c7d.out
   Power=
```
</details>

Parameters meet the criteria for the partition (memory and time). They otherwise would be pending indefinitely, as far as I could experience.
I tried removing all the parameters but the partition, attempted runs with and without specifying a workDirectory, using absolute and relative paths. The result is the same when specifying a number of thread. I always left sharedDirectory to default, but the `/home/user/.openmole/.tmp/ssh` is properly created so I guess it is not an issue.

Any thoughts on what I missed?

Cheers

**Comment by ScriBanana** [comment]
Update :
As far as I can tell, the `StdErr` file doesn't get written or at least not where the `scontrol` points, for some reason. The `WorkDir` doesn't even exist; it stops at the previous level (`/nfs/home/<user>/.openmole/.tmp/ssh/openmole-4d13481e-7466-4fb8-815d-e20d0f02b368/tmp/1639065600000` exists but is empty), even when the OM task is still running.
Also, I didn't mention that calculation are supposed to be ran from a directory on the Lustre bay of the cluster, where `~/scratch` leads (`scratch -> /lustre/<user>`). I did try `workDirectory = "/lustre/<user>"` without success. I'm able to run shell scripts with `sbatch` on it with no trouble, by the way.

**Comment by romainreuillon** [comment]
Can you make sure you can write in this directory from a job?

**Comment by ScriBanana** [comment]
To actually provide an answer : a job can definitely write on the lustre directory. At least when I run a sbatch on a script placed in this very directory.

**Comment by ScriBanana** [comment]
Alright, I get it. It indeed is not possible for a job to write anything on the head node, be it a log.
Now is there any work around that would have OM write the jobs outputs and errors on `workDirectory`, for example?

**Comment by romainreuillon** [comment]
Hello @ScriBanana, I've been quite buisy this week. Lets keep on this next week in visio.

**Comment by ScriBanana** [comment]
Hi Romain,

Not to worry, I was quite slow to making my way through the documentation available, to be honest.

Anyway I finally got though it! 😁
![image](https://user-images.githubusercontent.com/72523821/145612842-2983f78a-d432-481f-b339-ff024ba438d7.png)

Turns out there was a third bay with some directories I have access to, where jobs have write access, that is not listed on the cluster documentation for some reason, but that I was told about in an old forgotten email. Said email states that the `scratch` directory mentioned above is supposed to be writeable by jobs as well, by the way... I shall report that.

Have a nice weekend ! No visio this time, but maybe a beer someday, who knows ? :)
